desc/optimizable.py

```diff
@@ -0,0 +1,155 @@
+"""Base classes for optimizable objects."""
+
+import inspect
+import warnings
+from abc import ABC
+
+from desc.backend import jnp
+
+
+class Optimizable(ABC):
+    """Base class for all objects in DESC that can be optimized.
+
+    Sub-classes should decorate optimizable attributes with the
+    ``optimizable_parameter`` decorator.
+    """
+
+    @property
+    def optimizable_params(self):
+        """list: names of parameters that have been declared optimizable."""
```

Comment:
maybe specify in the docstring it is names as strings? I know it kind of obvious but I think better to be verbose